mono/metadata/icall.c

```diff
@@ -707,81 +697,54 @@ ves_icall_System_Array_CreateInstanceImpl64 (MonoReflectionType *type, MonoArray
 }
 
 ICALL_EXPORT gint32 
-ves_icall_System_Array_GetRank (MonoObject *arr)
+ves_icall_System_Array_GetRank (MonoObjectHandle arr, MonoError *error)
 {
-	return m_class_get_rank (mono_object_class (arr));
+	return m_class_get_rank (mono_handle_class (arr));
 }
 
+ICALL_EXPORT gint64
+ves_icall_System_Array_GetLongLength (MonoArrayHandle arr, gint32 dimension, MonoError *error);
+
 ICALL_EXPORT gint32
-ves_icall_System_Array_GetLength (MonoArray *arr, gint32 dimension)
+ves_icall_System_Array_GetLength (MonoArrayHandle arr, gint32 dimension, MonoError *error)
 {
-	gint32 rank = m_class_get_rank (mono_object_class (&arr->obj));
-	uintptr_t length;
-	ERROR_DECL (error);
-
-	if ((dimension < 0) || (dimension >= rank)) {
-		mono_error_set_index_out_of_range (error);
-		mono_error_set_pending_exception (error);
-		return 0;
-	}
-	
-	if (arr->bounds == NULL)
-		length = arr->max_length;
-	else
-		length = arr->bounds [dimension].length;
-
-#ifdef MONO_BIG_ARRAYS
+	uintptr_t const length = ves_icall_System_Array_GetLongLength (arr, dimension, error);
```

Comment:
(Just my personal preference) I don't really like this pattern (calling  an icall implementation from inside another function in the runtime).  I'd make a static helper and call it from both of the `ves_icall` functions.  I think there will be value in having icall implementations that are as mechanical as possible and delegate all the real work to helper functions.